### PR TITLE
[Arc] Add state lowering pass

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add the `LowerState` pass and accompanying operations to convert a design from a pure state transfer graph composed of arcs to a more procedural read-modify-write representation. After this transformation the design is a significant step down the path of becoming a software model.

The `LowerState` pass proceeds as follows:

- Group all `arc.state` ops according to their clock into `arc.clock_tree`s. Operations that are on direct input-to-output passthrough or on state-to-output paths are grouped into a `arc.passthrough` op.
- Introduce an explicit state storage allocation op for every state op, memory op, as well as every input and output port of the root module.
- Replace the root `hw.module` with a `arc.model` which no longer has any input and output ports, but a storage pointer argument instead. Storage allocation ops represent specific chunks of memory behind this pointer.
- Split every `arc.state` op with latency >0 up into a `arc.state_read`, `arc.state` with latency 0, and `arc.state_write` operation. This essentially breaks `arc.state` up into two parts: a read for all users of the state, and a transfer function plus write for all operands of the arc.

In doing so, the sea-of-gates representation of the HW dialect, which is a pure graph without op ordering constraints, is converted into a a sea-of-clocks, where each group contains the parts of the circuit that are triggered by the same clock. The actual computation that occurs on that trigger is represented procedurally in a proper SSA/CFG block.

TL;DR: This goes from "How do the gates connect together?" to "How does a computer simulate these gates?".

### CODE REVIEW

File: lib/Dialect/Arc/Transforms/LowerState.cpp [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+    if (!read) {
+      OpBuilder builder(op);
+      read = builder.create<StateReadOp>(value.getLoc(), state);
+    } else {
+      // We have no guarantee that we iterate over the uses in the order
+      // they appear in the block. Therefore make sure we move the read op
+      // before the earliest use.
+      auto *readOp = read.getDefiningOp();
+      if (op->isBeforeInBlock(readOp))
+        readOp->moveBefore(op);
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/LowerState.cpp:269]: I did a few (imprecise) measurements on compile time (used rocketConfig and largeBoomConfig from chipyard) of the lower state pass, both in current state and with these two lines removed and the insertion point of the builder always at start of block:

| | current [sec] | modified [sec] | speedup |
|---|---|---|---|
| rocket | ~1.8 | ~0.05 | 36x |
| boom | ~`420 | ~0.3 | 1400x |

I think the reason is that `isBeforeInBlock` has to recompute the ordering every iteration such that its runtime is not O(1) on average as mentioned in its documentation, but the full O(n) where n is the number of ops in the block. Since the body of the module is very big (because it contains all the arc calls, mem ops, etc.) we can clearly see the more than quadratic runtime.

File: include/circt/Dialect/Arc/Ops.td [hunk 2]
```diff
@@ -243,6 +243,133 @@ def MemoryWriteOp : ArcOp<"memory_write", [
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Trigger Grouping
+//===----------------------------------------------------------------------===//
+
+def ClockTreeOp : ArcOp<"clock_tree", [NoTerminator]> {
```
[the remark this answers is not in the document]
Comment (maerhart) [reply include/circt/Dialect/Arc/Ops.td:252]: Just figured out that there is a `NoRegionArguments` trait. Adding that should do the trick 🙂

[the remark this answers is not in the document]
Comment (fabianschuiki) [reply include/circt/Dialect/Arc/Ops.td:252]: Sweet, thanks for already looking up the solution :grin: :tada:

File: include/circt/Dialect/Arc/Ops.td [hunk 3]
```diff
@@ -253,6 +380,24 @@ def TapOp : ArcOp<"tap"> {
   let assemblyFormat = [{ $value attr-dict `:` type($value) }];
 }
 
+def ModelOp : ArcOp<"model", [RegionKindInterface, IsolatedFromAbove,
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply include/circt/Dialect/Arc/Ops.td:385]: Yeah that's a great idea. It really is supposed to only ever have that state pointer argument. At some later point we should just add a custom assembly format. I'll add the verifier for now.

File: lib/Dialect/Arc/Transforms/LowerState.cpp (hunk 1 above)
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/LowerState.cpp:269]: Fixed this by creating only one read op initially, and then doing a post-pass where I establish an op order separately and then sink all reads into the blocks where they are used. Could also have just created one read for every use, but that seemed a bit excessive.

File: lib/Dialect/Arc/Transforms/LowerState.cpp [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+  // the clock gates in the arc's clock tree.
+  auto info = getOrCreateClockLowering(stateOp.getClock());
+  info.enable = info.clock.getOrCreateAnd(info.enable, stateOp.getEnable(),
+                                          stateOp.getLoc());
+
+  // Allocate the necessary state within the model.
+  SmallVector<Value> allocatedStates;
+  for (unsigned stateIdx = 0; stateIdx < stateOp.getNumResults(); ++stateIdx) {
+    auto stateType = StateType::get(
+        stateOp.getResult(stateIdx).getType().cast<IntegerType>());
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/LowerState.cpp:355]: The `arc.state` ops results are not verified to be integers, so it might also be an array and crash here.

File: lib/Dialect/Arc/Transforms/LowerState.cpp [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+          continue;
+        }
+        mapped = operand;
+      }
+      mappedOperands.push_back(mapped);
+    } else {
+      // TODO: Try not to clone operations. Maybe we can drop the use references
+      // of the op on the worklist early, such that its dependencies already see
+      // the op as moved.
+      Operation *newOp = op->cloneWithoutRegions();
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/LowerState.cpp:175]: This should really be using `IRMapping` and the `clone` calls that also copy over nested regions.

File: lib/Dialect/Arc/Transforms/LowerState.cpp [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+  if (!stateOp.getClock()) {
+    stateOp.emitError("state with latency > 0 requires a clock");
+    return failure();
+  }
+
+  // Get the clock tree and enable condition for this state's clock. If this arc
+  // carries an explicit enable condition, fold that into the enable provided by
+  // the clock gates in the arc's clock tree.
+  auto info = getOrCreateClockLowering(stateOp.getClock());
+  info.enable = info.clock.getOrCreateAnd(info.enable, stateOp.getEnable(),
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/LowerState.cpp:348]: Good point. I'll just add the materialization here as well and then we can sort this out in the other PR>

File: lib/Dialect/Arc/Transforms/LowerState.cpp (hunk 4 above)
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/LowerState.cpp:355]: Good point. I've changed this to emit an error for now, but we'll probably want to change this to be able to support other types as well. Or have a pre-pass that lowers arc uses to trivial integers.
